Feature player search bar

Nilint1 opened the pull request:
Add a search bar and dropdown component to decide the search type.

Commit:
update .gitignore to ignore .vscode folder

Commit:
Initial change to add search.
Still a dirty search bar.  Needs refinement

Commit:
removed unneeded dropdown logic

Commit:
Add ability to search by position
Needed to add more filtering mechanisms.
Also added a SearchSelector to decide the term
to search on.

Commit:
Simplify the filter logic into one function
Problem:
4 filter functions is not KISS
Solution:
Combine filter functions into one function
that accepts and returns all types of inputs
and search types.

Commit:
Change SearchTerm->SearchType for clarity

Commit:
Changed ID attribute to number type
Problem:
String ID type is confusing comparted to other types
Solution:
Change ID to number type

Commit:
correct type error
Was converting a number to string needlessly

Commit:
Add additional containers to help control CSS

Commit:
Simplify data types for search types
Problem:
Unneeded id and interface present
Solution:
Remove the custom interface and revert to
just strings.  This supports KISS principals
and reduces unneeded complexity.

Commit:
hide the search element to allow for others to fix CSS.

Commit:
correct EOL characters

Nilint1 commented on .gitignore:
```diff
@@ -28,3 +28,6 @@ yarn-error.log*
 .env.production.local
+
+# ides
+.vscode
```
i have projects where I use 4 spaces per tab.  Had to tell VSCode this project only uses 2.  So, I ignore the .vscode folder.

Nilint1 commented on components/SearchBar/SearchBar.tsx:
```diff
@@ -0,0 +1,23 @@
+import styled from 'styled-components';
```
handles just the search bar logic

Nilint1 commented on components/SearchBar/searchBarFilterPlayers.tsx:
```diff
@@ -0,0 +1,46 @@
+import { Player, Goalie, PlayerRatings, GoalieRatings } from '../..'
```
Handles the actual filtering based on the search bar

Nilint1 commented on components/SearchBar/searchBarFilterPlayers.tsx:
```diff
@@ -0,0 +1,46 @@
+  }: Props): Filtered => {
+
+  // if nothing to filter, return the originals
+  if (!text || text == '' || !inputtedPlayers || !inputtedGoalies) {
```
if no search text, don't do any filtering

esilverm replied:
```suggestion
  if (!text || text === '' || !inputtedPlayers || !inputtedGoalies) {
```

Nilint1 commented on components/SearchBar/searchBarFilterPlayers.tsx:
```diff
@@ -0,0 +1,46 @@
+    };
+  }
+
+  const filteredPlayers = searchType.toLowerCase() == 'name' ? inputtedPlayers.filter((player) => {
```
based off the searchType, filter the objects

esilverm replied:
```suggestion
  const filteredPlayers = searchType.toLowerCase() === 'name' ? inputtedPlayers.filter((player) => {
```

Nilint1 commented on components/Selector/SearchSelector.tsx:
```diff
@@ -0,0 +1,78 @@
+import React, { useEffect, useRef, useState } from 'react';
```
Handles the Selector that decides what to search on.

Nilint1 commented on pages/[league]/players.tsx:
```diff
@@ -41,28 +44,56 @@ function PlayerPage({ league }: Props): JSX.Element {
     isLoading: isLoadingGoalies,
   } = useGoalieRatings(league);
 
+  // search functions
```
add logic to handle searching on multiple field types

Nilint1 commented on pages/[league]/players.tsx:
```diff
@@ -41,28 +44,56 @@ function PlayerPage({ league }: Props): JSX.Element {
+    })
+
   const getSkaters = () =>
-    skaterratings
```
replace the existing objects with the filtered versions.  If there is nothing the search for, the filtered objects match the original objects. i.e. no filtering

esilverm replied:
React Table includes a `useFilter` hook which may make this a bit more optimized. The problem with this is that it needs to re-render the whole table every time something is typed, so we may want to try to go with their solution instead for the rendering of the data. 

https://react-table.tanstack.com/docs/api/useFilters

Nilint1 replied:
okay, I will change it to use that hook.  This will take some time as it is kind of a complete rework.  Not a big deal, this is how I learn.

Nilint1 commented on pages/[league]/players.tsx:
```diff
@@ -81,6 +112,14 @@ function PlayerPage({ league }: Props): JSX.Element {
       />
       <Header league={league} activePage="players" />
       <Container>
+        <SearchWrapper>
```
This is where I was adding the search bar and selector to the page

Nilint1 commented on pages/[league]/players.tsx:
```diff
@@ -247,3 +286,22 @@ const SelectorWrapper = styled.div`
+  float: right;
+  padding-top: 10px;
+  padding-bottom: 10px;
+  display:none;
```
I am hiding the whole search function for now, here.

esilverm commented on components/SearchBar/searchBarFilterPlayers.tsx:
```diff
@@ -0,0 +1,46 @@
+    }) : inputtedPlayers.filter((player) => { return player.position.toLowerCase().includes(text.toLowerCase())
+    });
+
+  const filteredGoalies = searchType.toLowerCase() == 'name' ? inputtedGoalies.filter((player) => {
```
```suggestion
  const filteredGoalies = searchType.toLowerCase() === 'name' ? inputtedGoalies.filter((player) => {
```

Commit:
== -> ===

Need to use proper equal signs for logic statements.

Co-authored-by: Evan Silverman <es4753@nyu.edu>

Commit:
Revert players page back to original
Need to use React useFilter, so removing
existing filter from players page.

Commit:
Implement useFilters player filter
Uses a new SearchType data type to map a user friendly name
to an id to use for filtering.

Commit:
Add starting material for CSS.

Commit:
Correct filtering mechanism
Problem:
- Changing the search type did not call a filter
- typing a search term was updating on character behind
Solution:
- call filter on update of type
- search on event target value for typing in text

Commit:
correct CSS to be a little more clean

Nilint1 commented on components/ScoreTable/index.tsx:
```diff
@@ -1,10 +1,10 @@
-import React, { useMemo } from 'react';
-import { useTable, useSortBy, usePagination } from 'react-table';
+import React, { useMemo, useState } from 'react';
+import { useTable, useSortBy, usePagination, useFilters } from 'react-table';
```
now using filters from react-table

Nilint1 commented on components/ScoreTable/index.tsx:
```diff
@@ -1,10 +1,10 @@
-import { Player, Goalie } from '../..';
-
+import { Player, Goalie, SearchType } from '../..';
+import SearchBar from '../SearchBar/SearchBar'
```
using a search type interface to specify objects with human readable text and the id for the react-table

Nilint1 commented on components/ScoreTable/index.tsx:
```diff
@@ -88,14 +89,47 @@ Props): JSX.Element {
+  const [ searchType, setSearchType ] = useState(searchTypes[0].id);
+  const [ searchText, setSearchText ] = useState('');
+
+  const updateFilter = (text) => {
```
work to actually filter on the table

Nilint1 commented on components/ScoreTable/index.tsx:
```diff
@@ -88,14 +89,47 @@ Props): JSX.Element {
   return (
     <>
       {!hasData && <Notice>No results found</Notice>}
+      <SearchBar searchTypeOnChange={updateSearchType} searchTextOnChange={updateSearchText} searchTypes={searchTypes} />
```
grab all the onChange functions to do work on changes

Commit:
Add filter support for ratings table

Commit:
Remove the position search type from goalie tables

Nilint1 commented on components/ScoreTable/index.tsx:
```diff
@@ -88,14 +89,52 @@ Props): JSX.Element {
 
+  // search logic
+  // no need for position for goalies
+  const searchTypes: Array<SearchType> = players[0] && 'wins' in players[0] ? [
```
logic to remove the position search type if on a goalie table

Nilint1 commented on components/SearchBar/SearchBar.tsx:
```diff
@@ -0,0 +1,36 @@
+const SearchBarWrapper = styled.div`
+  font-size: 20px;
+  font-weight: 700;
+  display: none;
```
this is where I hide the search boxes

esilverm commented on components/RatingsTable/index.tsx:
```diff
@@ -85,14 +87,53 @@ Props): JSX.Element {
+  const [ searchText, setSearchText ] = useState('');
+
+  const updateFilter = (text) => {
+    console.log(allColumns);
```
this isnt necessary anymore

Nilint1 replied:
you are correct.  I missed one.  updated

esilverm reviewed (commented):
Just remove that log line and we are all good to merge and start styling

Commit:
remove unneeded logging message
